Create PyTorch DDP logging APIs for applications to use

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51260 add more logging fields that can be set in construction time
* **#50637 Create PyTorch DDP logging APIs for applications to use**

add APIs for logging pytorch ddp logging data in applications.

Differential Revision: [D25933411](https://our.internmc.facebook.com/intern/diff/D25933411/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25933411/)!

zhaojuanmao committed:
create a pytorch ddp detailed usage logger for analysis and debug

add APIs for logging pytorch ddp logging data in applications.

Differential Revision: [D25933411](https://our.internmc.facebook.com/intern/diff/D25933411/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25933411/)!

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3150,7 +3150,7 @@ def test_ddp_logging_data(self):
             self.assertEqual(ddp_logging_data.backend, 'gloo')
             self.assertEqual(ddp_logging_data.module_name, 'Net')
-            self.assertEqual(ddp_logging_data.device_ids, [])
+            self.assertEqual(ddp_logging_data.device_ids, "")
```
Is this empty because DDP is on a CPU device? Could we test it with GPU device and ensure that they are appropriately logged? 

Also, if it is CPU device looks like below output_device is -1 which is also CPU device. Can we make this consistent?

zhaojuanmao replied:
you mean make "output_device" as string, and output it as "" when they are none. That sounds good to me if they result in confusion.

zhaojuanmao committed:
Update on "Create PyTorch DDP logging APIs for applications to use"


add APIs for logging pytorch ddp logging data in applications.

Differential Revision: [D25933411](https://our.internmc.facebook.com/intern/diff/D25933411/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25933411/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Create PyTorch DDP logging APIs for applications to use"


add APIs for logging pytorch ddp logging data in applications.

Differential Revision: [D25933411](https://our.internmc.facebook.com/intern/diff/D25933411/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25933411/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Create PyTorch DDP logging APIs for applications to use"


add APIs for logging pytorch ddp logging data in applications.

Differential Revision: [D25933411](https://our.internmc.facebook.com/intern/diff/D25933411/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25933411/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Create PyTorch DDP logging APIs for applications to use"


add APIs for logging pytorch ddp logging data in applications.

Differential Revision: [D25933411](https://our.internmc.facebook.com/intern/diff/D25933411/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25933411/)!

[ghstack-poisoned]
